src/meta/src/stream/stream_manager.rs

```diff
@@ -288,26 +292,45 @@ where
             .await?;
 
         let mut env = Env {
-            upstream_parallel_unit_info,
+            upstream_fragment_vnode_info,
+            upstream_vnode_bitmap_info,
             tables_worker_actors,
             locations,
             dispatchers,
+            actor_vnode_bitmaps: &mut Default::default(),
             upstream_worker_actors,
         };
 
         for fragment in table_fragments.fragments.values_mut() {
+            if !fragment_upstream_table_map.contains_key(&fragment.fragment_id) {
+                continue;
+            }
+
             let is_singleton =
                 fragment.get_distribution_type()? == FragmentDistributionType::Single;
 
-            for actor in &mut fragment.actors {
+            for (idx, actor) in &mut fragment.actors.iter_mut().enumerate() {
                 let stream_node = actor.nodes.as_mut().unwrap();
                 env.resolve_chain_node_inner(
                     stream_node,
                     actor.actor_id,
+                    idx,
                     actor.same_worker_node_as_upstream,
                     is_singleton,
                 )?;
+                // setup actor vnode bitmap.
+                actor.vnode_bitmap = env.actor_vnode_bitmaps.remove(&actor.actor_id).unwrap();
```

Comment:
Can we directly pass `actor` in, so that the `vnode_bitmap` can be directly set in it? 🤔

Reply:
Actually not, `resolve_chain_node_inner` is called recursively and we also need a mutable reference of `StreamNode`.